controllers/spiaccesstoken_controller.go

```diff
@@ -166,15 +166,49 @@ func (r *SPIAccessTokenReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
-	// cleanup tokens which are in awaiting state and have their parent binging removed.
-	// grace period taken in account to avoid race condition during the object creation process (i.e. when token is created but not yet linked)
-	if at.Status.Phase == api.SPIAccessTokenPhaseAwaitingTokenData && time.Since(at.CreationTimestamp.Time).Seconds() > NoLinkingBindingGracePeriodSeconds {
+	tokenLifetime := time.Since(at.CreationTimestamp.Time).Seconds()
+
+	// to avoid double computing as it is an expensive operation
+	var hasBindings *bool
+
+	// cleanup tokens by lifetime
+	if tokenLifetime > r.Configuration.AccessTokenTtl.Seconds() {
 		hasLinkedBindings, err := hasLinkedBindings(ctx, &at, r.Client)
 		if err != nil {
 			lg.Error(err, "failed to validate the object", "token", at.ObjectMeta.Name, "error", err)
 			return ctrl.Result{}, fmt.Errorf("failed to validate the object: %w", err)
 		}
 		if !hasLinkedBindings {
```

Comment:
so what happens if token has some bindings? when it will be deleted? Shouldn't we schedule reconcile for this binding? https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L33

Reply:
I think this will just wait until finalizers give it a clear go when all the bindings are removed. This reconciler should react on the changes in the bindings, too..

So in another words, the `Delete()` will set the deletion timestamp and we will remove the finalizers once there are no more bindings linking to the token at which point it will finally be deleted...

Reply:
So, 
1) Is this code correct?
2) Should i explicitly call deletion from token storage, or it can be done by the finalizer?